presto-main/src/main/java/io/prestosql/execution/SqlTask.java

```diff
@@ -177,6 +185,9 @@ private void initialize(Function<SqlTask, ?> onDone, CounterStat failedTasks)
             catch (Exception e) {
                 log.warn(e, "Error running task cleanup callback %s", SqlTask.this.taskId);
             }
+
+            // notify that task is finished
+            notifyStatusChanged();
```

Comment:
do we not want to send the notification in case the following block (L160-170) returns from the listener? 
```            // store final task info
            while (true) {
                TaskHolder taskHolder = taskHolderReference.get();
                if (taskHolder.isFinished()) {
                    // another concurrent worker already set the final state
                    return;
                }
                if (taskHolderReference.compareAndSet(taskHolder, new TaskHolder(createTaskInfo(taskHolder), taskHolder.getIoStats()))) {
                    break;
                }
            }
```
Although, I don't know if the `return` block is possible during runtime.

Reply:
If
```
                    // another concurrent worker already set the final state
                    return;
```
was run it means that another thread set final final task info. This listener is the only place where final task info is set, so another thread must have called `notifyStatusChanged`

Reply:
do you mean the listener could get called twice from separate threads? is that possible? - and if it does then the `failedTasks` counter computed might be wrong.

Reply:
> do you mean the listener could get called twice from separate threads? is that possible? - and if it does then the failedTasks counter computed might be wrong.

That's a good point. I don't think it can be called twice for done state. I guess author of this code wanted to make sure that `compareAndSet` is use in proper context regardless.

Reply:
If you look at the code it's quite old 2014 and `failedTasks` code was added much later. I guess the original concurrency issue might not be there anymore.

Reply:
Yes, you are right. It shouldn't exist anymore.